src/regex.cr

```diff
@@ -530,6 +530,21 @@ class Regex
     match(str, pos, options: _options)
   end
 
+  # Matches a regular expression against `String` *str*. This starts at the character
+  # index *pos* if given, otherwise at the start of *str*. Returns a `Regex::MatchData`
+  # if *str* matched, otherwise raises an exception. `$~` will contain the same value
+  # if matched.
+  #
+  # ```
+  # /(.)(.)(.)/.match!("abc")[2]        # => "b"
+  # /(.)(.)/.match!("abc", 1)[2]        # => "c"
+  # /(.)(タ)/.match!("クリスタル", 3)[2] # raises Exception
+  # ```
+  def match!(str : String, pos : Int32 = 0, *, options : Regex::MatchOptions = :none) : MatchData
+    byte_index = str.char_index_to_byte_index(pos) || raise Error.new "no matches found"
+    $~ = match_at_byte_index(str, byte_index, options) || raise Error.new "no matches found"
```

Comment:
Use sentence case for error messages.


```suggestion
    byte_index = str.char_index_to_byte_index(pos) || raise Error.new "No matches found"
    $~ = match_at_byte_index(str, byte_index, options) || raise Error.new "No matches found"
```

Reply:
Actually shouldn't this read `Match not found` - since the _match_ in this case IIUC means the whole expression?

Reply:
As far as I'm aware, there isn't a (consistent) convention for error message casing so I don't think that's relevant here (#12031).

I believe that's up to interpretation, but I'm not opposed to using "Match not found" instead.

Reply:
Lower cased error messages are being thrown by the compiler. Stdlib used sentence cased ones.